man/coap_encryption.txt.in

```diff
@@ -0,0 +1,641 @@
+as detect what is the version of the currently loaded TLS library.
+
+*NOTE:* If OpenSSL is being used, then the minimum supported OpenSSL library
+version is 1.1.0.
+
+Network traffic can be encrypted or un-encrypted with libcoap if there is an
+underlying TLS library.
+
+If TLS is going to be used for encrypting the network traffic, then the TLS
+information for Pre-Shared Keys (PSK) or Public Key Infrastructure (PKI) needs
+to be configured before any network traffic starts to flow. For Servers, this
+has to be done before the Endpoint is created, for Clients, this is done
+during the Client Session set up.
+
+For Servers, all the encryption information is held internally by the TLS
+Context level and the CoAP Context level as the Server is listening for new
+incoming traffic based on the Endpoint definition.  The TLS and CoAP session
+will not get built until the new traffic starts, which is done by the libcoap
+library, with the session having a reference count of 1.
+
+For Clients, all the encryption information can be held at the TLS Context and
+CoAP Context levels, or at the TLS Session and CoAP Session levels.  If
+defined at the Context level, then when Sessions are created, they will
+inherit the Context definitions, unless they have separately been defined for
+the Session level, in which case the Session version will get used.
+Typically the information will be configured at the Session level for Clients.
+
+In principle the set-up sequence for CoAP Servers looks like
+----
+coap_new_context()
+coap_context_set_pki_root_cas() - if the root CAs need to be updated and PKI
+coap_context_set_pki() and/or coap_context_set_psk() - if encryption is required
+coap_new_endpoint()
+----
+
+Multiple endpoints can be set up per Context, each listening for a new traffic
+flow with different TCP/UDP protocols, TLS protocols, port numbers etc. When a
+new traffic flow is started, then the CoAP library will create and start a new
+server session.
+
+In principle the set-up sequence for CoAP Clients looks like
+----
+coap_new_context()
+coap_context_set_pki_root_cas() - if the root CAs need to be updated and PKI
+coap_new_client_session(), coap_new_client_session_pki() or coap_new_client_session_psk()
+----
+
+Multiple client sessions are supported per Context.
+
+Due to the nature of TLS, there are Callbacks that are invoked as the TLS
+session negotiates encryption algorithms, encryption keys etc.
+Where possible, the CoAP layer handles all this automatically based on
+different configuration options passed in by the coap_*_pki() functions.
+
+For PSK setup, the required information needs to be provided in the setup
+calls with no application Callbacks required. Both the Client and Server have
+to provide a PSK.  The Server must have a Hint defined and the Client must
+have an Identity defined.
+
+For PKI setup, if the libcoap PKI configuration options do not handle a specific
```

Comment:
I'm not entirely sure after investigation that OpenSSL 1.1.0 properly supports RPK, but am willing to be proven wrong.  Does tinyDTLS support this as an alternative?

Reply:
Yes, TinyDTLS supports RPK (and only RPK, no full-fledged PKI).

Reply:
The reason you might want asymmetric cryptography but not check certificate chains (consider 150 or so root CAs stored in a class 1 IoT device...)

Reply:
Fully understand why RPK makes sense, need it for the DOTS protocol.  However https://en.wikipedia.org/wiki/Comparison_of_TLS_implementations confirms my suspicion of OpenSSL as unknown.  My guess is that RPK would be done using the asn1 variant.

Reply:
I was specifically concerned about the API. I would expect to use RPKs in libcoap similar to full X.509 certificates but without transitive trust. The root CA files then pretty much would be the known RPKs, and the certfiles an RPK instead of a full certificate. (All being DER-encoded ASN.1 structures.)

Reply:
Okay, sounds reasonable to me. I suppose, this would be somehow related to PR #128.

Reply:
Yes, I was planning to use what made sense out of #128 with these new functions for TinyDTLS.